pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1507,24 +1556,42 @@ public CompletableFuture<Void> close(
         CompletableFuture<Void> closeFuture = new CompletableFuture<>();
 
         lock.writeLock().lock();
+        CompletableFuture<Void> inProgressTransferCloseTask = null;
         try {
             if (!disconnectClients) {
                 transferring = true;
             }
             // closing managed-ledger waits until all producers/consumers/replicators get closed. Sometimes, broker
             // forcefully wants to close managed-ledger without waiting all resources to be closed.
-            if (!isClosingOrDeleting || closeWithoutWaitingClientDisconnect) {
-                fenceTopicToCloseOrDelete();
+            if (isClosingOrDeleting) {
+                if (!disconnectClients) {
+                   return closeFutures.notDisconnectClients;
+                }
+                /** There is a in-progress half closing task. see the section 2-b-1 of {@link CloseFutures}. **/
+                if (transferring && closeFutures.notWaitDisconnectClients == null) {
+                    inProgressTransferCloseTask = closeFutures.notDisconnectClients;
+                } else {
+                    if (closeWithoutWaitingClientDisconnect) {
+                        return closeFutures.notWaitDisconnectClients;
+                    }
+                    return closeFutures.waitDisconnectClients;
+                }
+            }
+            fenceTopicToCloseOrDelete();
```

Comment:
nit:  set inProgressTransferCloseTask after dedup

```java
/** There is a in-progress half closing task. see the section 2-b-1 of {@link CloseFutures}. **/
if (transferring && disconnectClients) {
  inProgressTransferCloseTask = closeFutures.transferring
}
```

Reply:
Since we should not combine the `notWaitDisconnectClients` and `waitDisconnectClients.` Ignored this suggestion.